### ISSUE TITLE
S4035 causes confusion and it looks like it gives us false positives.

### ISSUE THREAD

**Issue by ndsvw** [SonarSource/sonar-dotnet#5101]
### Description

This is about rule S4035.

We are using version 8.9.1 of SonarQube, but it does not look like you changed something related to this problem in the newer versions.

Rule S4035 is "Seal class 'AbstractBase' or implement 'IEqualityComparer<T>' instead".
If I click "Why is this an issue", it tells me "Classes implementing 'IEquatable<T> should be sealed".

But in our case, it does not make sense for multiple reasons.

### Repro steps

```c#
public abstract class AbstractBase{
    public override bool Equals(object obj){
        //calls the private Equals method in the end.
    }

    private bool Equals(AbstractBase other){
        // return true oder false
    }

    public override int GetHasCode(){
        // returns hash code
    }
}
```

```c#
public class Derived : AbstractBase{
    public override bool Equals(object obj){
        //calls the private Equals method in the end.
    }

    private bool Equals(Derived other){
        // return base.Equals(other) && .....
    }

    public override int GetHasCode(){
        // returns hash code
    }
}
```

### Expected behavior

No issues.

### Actual behavior

For both my classes, I get a S4035 warning.
Either it does not make sense or the descriptions of the issue does not explain it good enough.

1) Both classes do not implement the IEquatable<T> interface. Regardless, the S4035 title says "Classes implementing 'IEquatable<T> should be sealed". I do not implement it. So there should be no reason for this warning.

2) My first class is abstract. The warning: "Seal class 'AbstractBase' or implement 'IEqualityComparer<T>' instead". It is not even possible to seal an abstract class because it does not make sense.

3) I've seen [this issue](https://github.com/SonarSource/sonar-dotnet/issues/4000) that is similar to mine, but the argument of pavel does not seem to be true as my own Equals method is not protected. It's private.

### Related information

* MSBuild version: 16.11.0
* Operating System: Windows
* SonarScanner for MsBuild version: 5.3.2

**Comment by pavel-mikula-sonarsource** [comment]
Hi @ndsvw,

Thank you for reporting this. I confirm it as a False Positive.

**Comment by pavel-mikula-sonarsource** [comment]
_Internal note:_ We need to check what was this piece originally meant to do and either fix it or change the message in this scenario:
https://github.com/SonarSource/sonar-dotnet/blob/ffd12eacece182259b10b9642b9453255d9d62fa/analyzers/src/SonarAnalyzer.CSharp/Rules/EquatableClassShouldBeSealed.cs#L85-L87

**Comment by andrei-epure-sonarsource** [comment]
Hello @ndsvw 

Indeed the message is confusing.

Coming back to your example, for me to understand the usecase: is there a particular reason why you are not implementing `IEquatable`?
